File: Source Code/Python/heapsort.py
import time
start_time = time.time()
arr = []
def heapify(n, i):
    largest = i
    l = 2 * i + 1 
    r = 2 * i + 2
 
    if l < n and arr[i] < arr[l]:
        largest = l
    if r < n and arr[largest] < arr[r]:
        largest = r
    if largest != i:
        (arr[i], arr[largest]) = (arr[largest], arr[i])  # swap
        heapify(n, largest)
 
n = 0
def heapSort():

 
    for i in range(n // 2, -1, -1):
        heapify(n, i)
 
 
    for i in range(n - 1, 0, -1):
        (arr[i], arr[0]) = (arr[0], arr[i])  # swap
        heapify(i, 0)

filename = "input.inp"
with open('input.inp') as f:
    n = int(f.readline())
    line = f.readline()
    data = line.split()
    for i in range(n):
        arr.append(int(data[i]))
heapSort()

File: Source Code/C++/heapsort.cpp
#include <bits/stdc++.h>
using namespace std;
vector<int> arr;
void heapify(int n, int i){

    int largest = i;
    int l = 2 * i + 1;
    int r = 2 * i + 2;
    if (l < n && arr[l] > arr[largest])
        largest = l;
    if (r < n && arr[r] > arr[largest])
        largest = r;
    if (largest != i) {
        swap(arr[i], arr[largest]);
        heapify(n, largest);
    }
}
void heapSort(){
    int n = arr.size();
    for (int i = n / 2 - 1; i >= 0; i--)
        heapify(n, i);
    for (int i = n - 1; i > 0; i--) {
        swap(arr[0], arr[i]);
        heapify(i, 0);
    }
}
int main()
{
    freopen("input.inp", "r", stdin);
    int n;
    cin >> n;
    arr.resize(n);
    for (int i = 0; i < n; i++)
        cin >> arr[i];
    heapSort();
    return 0;
}